src/librustc_mir/transform/const_prop.rs

```diff
@@ -969,13 +970,40 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             | TerminatorKind::GeneratorDrop
             | TerminatorKind::FalseEdges { .. }
             | TerminatorKind::FalseUnwind { .. } => {}
-            //FIXME(wesleywiser) Call does have Operands that could be const-propagated
-            TerminatorKind::Call { .. } => {}
+            TerminatorKind::Call { ref mut args, .. } => {
+                let mir_opt_level = self.tcx.sess.opts.debugging_opts.mir_opt_level;
+                // Constant Propagation into function call arguments is gated
+                // under mir-opt-level 2, because LLVM codegen gives performance
+                // regressions with it.
+                if mir_opt_level >= 2 {
+                    // Every argument in our function calls can be const propagated.
+                    // We try to, before exiting.
+                    for opr in args {
+                        if let Some(l) = opr.place().and_then(|p| p.as_local()) {
+                            if let Some(value) = self.get_const(l) {
+                                if self.should_const_prop(value) {
+                                    if let interpret::Operand::Immediate(
+                                        interpret::Immediate::Scalar(
```

Comment:
Here's a brain dump about the scalar thing:

Currently, we only have a way to create consts from scalars. The representation of values in miri that we care about fall into a few cases:

- [`Operand`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/enum.Operand.html)
  - This enum is the most general representation of values in miri. It abstracts over both direct values and values help indirectly in memory.
  - This is the `Op` in `OpTy`.

- `Immediate`
  - This represents a value directly. It can be 0, 1, or 2 individual `Scalar`s. You're probably thinking "why two?", that's because checked math is very common and is represented as a tuple of `(over/under-flowed: bool, result: integer)` and having the values not spill into Indirect memory significantly improves performance. 

- `Indirect`
  - This represents all other values and is backed by an allocation somewhere in miri. This could either be because a user requested it (eg `let x = Box::new(4u8);`) or because the value doesn't fit in an `Immediate`. 

This code checks for the very common case of an argument that is expressed by one scalar but misses the case where it fits in two scalars. Unifying the treatment of `Operands` as @oli-obk mentioned in your original PR would be a good way to handle that without duplicating code.

Reply:
> This code checks for the very common case of an argument that is expressed by one scalar but misses the case where it fits in two scalars. Unifying the treatment of Operands as @oli-obk mentioned in your original PR would be a good way to handle that without duplicating code.

Indeed. I want to add that to my to-do list as well :)

I don't know if you've seen it, but [the code for propagating consts into rvalues](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/const_prop.rs#L613) includes a handler for `ScalarPair` cases. That's why the `tuple_literal_propagation` test does show some constant propagation: it does so [for the assignment statement](https://github.com/rust-lang/rust/blob/master/src/test/mir-opt/const_prop/tuple_literal_propagation/rustc.main.ConstProp.diff). Maybe we're able to take ideas from there to generalize the `Operand` treatment as @oli-obk suggested :)